keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,484 @@
+namespace: kube-system
+annotations:
+  scheduler.alpha.kubernetes.io/critical-pod: ''
+labels:
+  tier: node
+  component: kube-proxy
+  unrestricted.node.kubernetes.io/example: true
+```
+
+Might produce a mirror pod (on a node named `testing-default-pool-588c6e30-408q`) with metadata:
+
+```yaml
+name: kube-proxy-testing-default-pool-588c6e30-408q
+namespace: kube-system
+annotations:
+  kubernetes.io/config.hash: 5edeeae4fbd5dde04f6298cdd8d7fb15
+  kubernetes.io/config.mirror: 5edeeae4fbd5dde04f6298cdd8d7fb15
+  kubernetes.io/config.seen: "2019-10-10T07:38:16.378551617Z"
+  kubernetes.io/config.source: file
+  kubernetes.io/config.removed-labels: '{"component": "kube-proxy", "tier": "node"}'
+  scheduler.alpha.kubernetes.io/critical-pod: ""
+labels:
+  unrestricted.node.kubernetes.io/example: true
+creationTimestamp: "2019-10-10T07:38:21Z"
+resourceVersion: "29997713"
+selfLink: /api/v1/namespaces/kube-system/pods/kube-proxy-testing-default-pool-588c6e30-408q
+uid: ef19cb63-eb30-11e9-a0d3-42010a80005d
+ownerReferences:
+- apiVersion: v1
+  kind: Node
+  name: testing-default-pool-588c6e30-408q
+  uid: 40e7848c-a828-11e9-829c-42010a80016f
+  controller: true
+  blockOwnerDeletion: false
+```
+
+### Risks and Mitigations
+
+Some Kubernetes setups depend on statically serving services today. Applying these mitigations will
+likely break these clusters. There is no way to apply these changes in a fully backwards compatible
+way, so instead we will rely on a staged rollout through the `MirrorPodNodeRestriction` and
+`MirrorPodMetadataFilter` feature gates, and call out the actions required.
+
+Clusters currently depending on label-matching static pods will need to migrate the
+labels to the new whitelisted key prefix. This can be done in a non-disruptive way,
+but requires a multistep process. For example, to migrate static pods providing a Service:
+
+1. Update the static pods (by deploying updated static manifests) with _both_ the old & new labels.
+2. Update the service selector to match the new labels.
+3. Update the static pods to remove the old labels.
+
+## Design Details
+
+### Test Plan
+
+Our CI environment does not depend on static pods serving services, so we can enable the feature
+gate in the standard Kubernetes E2E environment. The restrictions can be verified by impersonating a
+node's identity and ensuring illegal mirror pods cannot be created.
+
+To test the Kubelet modifications, a privileged pod can write a static pod manifest to a node and
+verify the expected changes are made.
+
+### Graduation Criteria
+
+The feature gate will initially be in a default-disabled alpha state. Graduating to beta will make
+the feature enabled by default, but users that have not yet updated existing label
+usage to the unrestricted keys can still disable it. We will allow at least 2 releases before
+migrating to GA and removing the feature gate entirely.
+
+Here is an approximate graduation schedule, specific release numbers subject to change:
```

Comment:
I'm not sure if it belongs in the graduation schedule or in a separate schedule, but it would be helpful and informative to lay out the steps a deployer using labels on static pods for service routing would have to follow to safely transition pods to the new labels over N releases (assuming deployers do one config per Kubernetes version):

* Current state (~v1.16):
  * pods use label foo=true
  * services use selector foo=true
* Release N - deployer starts to make changes:
  * pods use labels foo=true,unrestricted.node.kubernetes.io/foo=true
  * services use selector foo=true (might have kubelets at N-2, N-1, or N)
* Release N+1:
  * pods use labels foo=true,unrestricted.node.kubernetes.io/foo=true
  * services use selector foo=true (might have kubelets at N-1, N, or N+1)
* Release N+2:
  * pods use labels unrestricted.node.kubernetes.io/foo=true
  * services use selector unrestricted.node.kubernetes.io/foo=true (oldest kubelet is at N)

A timeline like that would imply we should not enable kubelet rewriting until ~1.19.